### PR TITLE
fix(backend): fixed validation logic and resource manager logic when creating job and run

### LINKED ISSUE #4900: Discuss the validation and precendence of pipeline ID, workflow manifest and pipeline version
### What steps did you take:

1. Create a pipeline
2. Upload multiple versions to the same pipeline with different manifests (let's call the versions 1, 2, 3, and their respective manifests 1', 2', 3'. First created version is 1, official default version is 3)
3. Compile a pipeline in local and build the package with version 4 and its respective manifest is 4'

### What happened:
```
pipeline_id = "${pipeline_id}"
pipeline_package_path = "${path of version 4}"
version_2_id = "${version 2 id}"

# Run version 4
run_with_package_path_only = cl.run_pipeline(
    experiment_id, "run_with_package_path_only", pipeline_package_path=pipeline_package_path)

# Run version 3
run_with_pipeline_id_only = cl.run_pipeline(
    experiment_id, "run_with_pipeline_id_only", pipeline_id=pipeline_id)

# Run version 2
run_with_version_id_only = cl.run_pipeline(
    experiment_id, "run_with_version_id_only", version_id=version_2_id)

# Run version 2
run_with_version_id_and_package_path = cl.run_pipeline(
    experiment_id, "run_with_version_id_and_package_path", version_id=version_2_id,
    pipeline_package_path=pipeline_package_path)

# Run version 2
run_with_pipeline_id_and_version_id = cl.run_pipeline(
    experiment_id, "run_with_pipeline_id_and_version_id", version_id=version_2_id,
    pipeline_id=pipeline_id)

# Run version 2
run_with_pipeline_id_and_package_path_and_version_id = cl.run_pipeline(
    experiment_id, "run_with_pipeline_id_and_package_path_and_version_id", pipeline_id=pipeline_id,
    pipeline_package_path=pipeline_package_path, version_id=version_2_id)

# 400.
# Neither pipeline spec nor pipeline version is valid.
# Invalid input error: Please specify a pipeline version in Run's resource references:
# Invalid input error: Please either specify a pipeline ID or a workflow manifest, not both.
# This error is from server layer(validation logic)
try:
    run_with_pipeline_id_and_package_path = cl.run_pipeline(
        experiment_id, "run_with_pipeline_id_and_package_path", pipeline_id=pipeline_id,
        pipeline_package_path=pipeline_package_path)
except Exception as e:
    print("run run_with_pipeline_id_and_package_path encountered error {}".format(e))

# 404
# ResourceNotFoundError: WorkflowSpecManifest run_with_valid_pipeline_id_and_invalid_version_id  not found.
# Failed to fetch workflow spec manifest.
# The error is from Resource Manager layer
try:
    run_with_valid_pipeline_id_and_invalid_version_id = cl.run_pipeline(
        experiment_id, "run_with_valid_pipeline_id_and_invalid_version_id ", pipeline_id=pipeline_id,
        version_id=version_2_id + "invalid")
except Exception as e:
    print("run_with_valid_pipeline_id_and_invalid_version_id encountered error {}".format(e))

# 404 ResourceNotFoundError: Pipeline 9fce6aec-39a2-420b-88c9-a1f3e16d5f0dinvalid not found.
# This error came from Pipeline Store layer
try:
    run_with_invalid_pipeline_id_and_valid_version_id = cl.run_pipeline(
        experiment_id, "run_with_invalid_pipeline_id_and_valid_version_id ", pipeline_id=pipeline_id + "invalid",
        version_id=version_2_id)
except Exception as e:
    print("run_with_invalid_pipeline_id_and_valid_version_id encountered error {}".format(e))
```
I checked the code, the validation logic is here:
https://github.com/kubeflow/pipelines/blob/89e42105bd941e1d918d0bff9ba0c4f80b106ed4/backend/src/apiserver/server/run_server.go#L107
It will try to validate PipeLineSpec. If it encounters error when validating PipeLineSpec then it will try to validate PipelineVersionReference. It WON'T check PipelineVersionReference when PipeLineSpec is valid and it WON'T throw error when PipelineVersionReference  is valid and PipeLineSpec is invliad.

And the creation logic is here:
https://github.com/kubeflow/pipelines/blob/89e42105bd941e1d918d0bff9ba0c4f80b106ed4/backend/src/apiserver/resource/resource_manager.go#L317

For the precedence, pipeline id and workflow manifest has same priority.
version id has higher priority than pipeline id and workflow manifest

### What did you expect to happen:
1. When user specify both  pipeline_package_path(equal to pipeline_manifest) and  version_id, now the behavior is that version_id should take precedence. However, corner cases may happen. Based on the logic https://github.com/kubeflow/pipelines/blob/89e42105bd941e1d918d0bff9ba0c4f80b106ed4/backend/src/apiserver/resource/resource_manager.go#L317, if it encounters error in https://github.com/kubeflow/pipelines/blob/89e42105bd941e1d918d0bff9ba0c4f80b106ed4/backend/src/apiserver/resource/resource_manager.go#L327, it will try to  getWorkflowSpecBytesFromPipelineSpec. 
Not sure whether this fallback makes sense here. And not sure what pipeline version UI will show.

**I think when user can't set pipeline id or version id  when user set workflow manifest**

2.
Validation logic is confusing and incomplete
https://github.com/kubeflow/pipelines/blob/89e42105bd941e1d918d0bff9ba0c4f80b106ed4/backend/src/apiserver/server/run_server.go#L107 won't validate PipelineVersionReference when PipeLineSpec is valid.
For example, when user specify a valid pipeline id and an invalid version id, it will pass the validation logic and it will throw not found error when resource manager finds out the workflow.Workflow is nil https://github.com/kubeflow/pipelines/blob/89e42105bd941e1d918d0bff9ba0c4f80b106ed4/backend/src/apiserver/resource/resource_manager.go#L345.
The error msg is confusing here.  
**I think we should add validation logic when user specify any invalid input no matter whether it's pipeline id, version id or workflow manifest. 
If there is any invalid input, the validation layer will throw error and backend should not call resource manager layer or store layer**

The output should be as follows:
```
pipeline_id = "${pipeline_id}"
pipeline_package_path = "${path of version 4}"
version_2_id = "${version 2 id}"

# Run version 4
run_with_package_path_only = cl.run_pipeline(
    experiment_id, "run_with_package_path_only", pipeline_package_path=pipeline_package_path)

# Run version 3
run_with_pipeline_id_only = cl.run_pipeline(
    experiment_id, "run_with_pipeline_id_only", pipeline_id=pipeline_id)

# Run version 2
run_with_version_id_only = cl.run_pipeline(
    experiment_id, "run_with_version_id_only", version_id=version_2_id)

# invalid, 400, user can't set pipeline id or version id  when user set workflow manifest
run_with_version_id_and_package_path = cl.run_pipeline(
    experiment_id, "run_with_version_id_and_package_path", version_id=version_2_id,
    pipeline_package_path=pipeline_package_path)

# Run version 2
run_with_pipeline_id_and_version_id = cl.run_pipeline(
    experiment_id, "run_with_pipeline_id_and_version_id", version_id=version_2_id,
    pipeline_id=pipeline_id)

# invalid, 400, user can't set pipeline id or version id  when user set workflow manifest
run_with_pipeline_id_and_package_path_and_version_id = cl.run_pipeline(
    experiment_id, "run_with_pipeline_id_and_package_path_and_version_id", pipeline_id=pipeline_id,
    pipeline_package_path=pipeline_package_path, version_id=version_2_id)

# invalid, 400, user can't set pipeline id or version id  when user set workflow manifest
try:
    run_with_pipeline_id_and_package_path = cl.run_pipeline(
        experiment_id, "run_with_pipeline_id_and_package_path", pipeline_id=pipeline_id,
        pipeline_package_path=pipeline_package_path)
except Exception as e:
    print("run run_with_pipeline_id_and_package_path encountered error {}".format(e))

# invalid, 400, user can't specify invalid version id
try:
    run_with_valid_pipeline_id_and_invalid_version_id = cl.run_pipeline(
        experiment_id, "run_with_valid_pipeline_id_and_invalid_version_id ", pipeline_id=pipeline_id,
        version_id=version_2_id + "invalid")
except Exception as e:
    print("run_with_valid_pipeline_id_and_invalid_version_id encountered error {}".format(e))

# invalid, 400, user can't specify invalid pipeline id
try:
    run_with_invalid_pipeline_id_and_valid_version_id = cl.run_pipeline(
        experiment_id, "run_with_invalid_pipeline_id_and_valid_version_id ", pipeline_id=pipeline_id + "invalid",
        version_id=version_2_id)
except Exception as e:
    print("run_with_invalid_pipeline_id_and_valid_version_id encountered error {}".format(e))
```

 

### Environment:

How did you deploy Kubeflow Pipelines (KFP)?

KFP version:  1.1.2

KFP SDK version: .1.2

### Anything else you would like to add:
[Miscellaneous information that will assist in solving the issue.]

/kind bug

### PR DESCRIPTION
**Pull request by capri-xiyue**
**Description of your changes:**
Fixed #4900 
fixed validation logic and resource manager logic when creating job and run based on discussion of #4900
**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

File: backend/src/apiserver/server/job_server_test.go [hunk 1]
```diff
@@ -162,7 +157,60 @@ func TestValidateApiJob_NoValidPipelineSpecOrPipelineVersion(t *testing.T) {
 	}
 	err := server.validateCreateJobRequest(&api.CreateJobRequest{Job: apiJob})
 	assert.Equal(t, codes.InvalidArgument, err.(*util.UserError).ExternalStatusCode())
-	assert.Contains(t, err.Error(), "Neither pipeline spec nor pipeline version is valid")
+	assert.NotNil(t, err)
+	assert.Contains(t, err.Error(), "Please specify a pipeline by providing a pipeline spec or/and pipeline version.")
+}
+
+func TestValidateApiJob_WorkflowManifestAndPipelineVersion(t *testing.T) {
```
Comment (capri-xiyue) [inline backend/src/apiserver/server/job_server_test.go:164]: Not sure whether I need to duplicate the UT for job and run since the UT of such logic have been covered in UT of validatePipelineSpecAndCheckPipelineVersionReference

File: backend/src/apiserver/server/run_server_test.go [hunk 2]
```diff
@@ -455,6 +456,19 @@ func TestValidateCreateRunRequest_EmptyName(t *testing.T) {
 	assert.Contains(t, err.Error(), "The run name is empty")
 }
 
+func TestValidateCreateRunRequest_InvalidPipelineVersionReference(t *testing.T) {
```
Comment (capri-xiyue) [inline backend/src/apiserver/server/run_server_test.go:459]: Not sure whether I need to duplicate the UT for job and run since the UT of such logic have been covered in UT of validatePipelineSpecAndCheckPipelineVersionReference

File: backend/src/apiserver/server/util.go [hunk 3]
```diff
@@ -216,9 +216,25 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
...
+	}
+	if spec != nil && spec.GetWorkflowManifest() != "" && pipelineVersionId != "" {
+		return util.NewInvalidInputError("Please either specify a pipeline version or a workflow manifest, not both.")
+	}
+	if err := ValidatePipelineSpec(resourceManager, spec); err != nil {
+		return err
+	}
+	return nil
+}
 func ValidatePipelineSpec(resourceManager *resource.ResourceManager, spec *api.PipelineSpec) error {
```
Comment (capri-xiyue) [inline backend/src/apiserver/server/util.go:235]: Just wondering why `ValidatePipelineSpec` is public. Looks like it is only used in this package.

File: backend/src/apiserver/server/util.go [hunk 4]
```diff
@@ -253,27 +269,23 @@ func ValidatePipelineSpec(resourceManager *resource.ResourceManager, spec *api.P
 // (1) a pipeline version is specified in references as a creator.
 // (2) the above pipeline version does exists in pipeline version store and is
 // in ready status.
-func CheckPipelineVersionReference(resourceManager *resource.ResourceManager, references []*api.ResourceReference) (*string, error) {
```
Comment (capri-xiyue) [inline backend/src/apiserver/server/util.go:256]: Same here. Just wondering why `CheckPipelineVersionReference` is public. Looks like it is only used in this package.

File: backend/src/apiserver/resource/resource_manager_util.go [hunk 5]
```diff
@@ -245,7 +245,7 @@ func OverrideParameterWithSystemDefault(workflow util.Workflow, apiRun *api.Run)
 // This is for legacy usage of pipeline id to create run. The standard way to
 // create run is by specifying the pipeline version.
 func ConvertPipelineIdToDefaultPipelineVersion(pipelineSpec *api.PipelineSpec, resourceReferences *[]*api.ResourceReference, r *ResourceManager) error {
```
Comment (capri-xiyue) [inline backend/src/apiserver/resource/resource_manager_util.go:247]: Same here. Why this method is public?

File: backend/src/apiserver/server/util_test.go [hunk 6]
```diff
@@ -336,6 +335,92 @@ func TestValidatePipelineSpec_ParameterTooLong(t *testing.T) {
 	assert.Contains(t, err.Error(), "The input parameter length exceed maximum size")
 }
 
+func TestCheckPipelineVersionReference_PipelineVersionReference(t *testing.T) {
```
Comment (capri-xiyue) [inline backend/src/apiserver/server/util_test.go:338]: I know usually we don't do UT for private methods. However, if we don't do UT here, we will need to do duplicates UT for both job and run.

Comment (capri-xiyue) [inline backend/src/apiserver/server/util_test.go:338]: I know usually we don't do UT for private methods. However, if we don't do UT here, we will need to do duplicates UT for both job and run.

File: backend/src/apiserver/server/util.go [hunk 7]
```diff
@@ -216,9 +216,25 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
...
+	if (spec == nil || (spec.GetPipelineId() == "" && spec.GetWorkflowManifest() == "")) && pipelineVersionId == "" {
+		return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline spec or/and pipeline version.")
+	}
+	if spec != nil && spec.GetWorkflowManifest() != "" && pipelineVersionId != "" {
+		return util.NewInvalidInputError("Please either specify a pipeline version or a workflow manifest, not both.")
+	}
+	if err := ValidatePipelineSpec(resourceManager, spec); err != nil {
+		return err
+	}
+	return nil
```
Comment (Bobgy) [inline backend/src/apiserver/server/util.go:233]: nit: the if conditions seem a little hard to parse, what do you think about the following?
```suggestion
	// centralize all the nil checks here
	pipelineId := spec != nil ? spec.GetPipelineId() : ""
	workflowManifest := spec != nil ? spec.GetWorkflowManifest() : ""
	pipelineVersionId, err := CheckPipelineVersionReference(resourceManager, resourceReferences)
	if err != nil {
		return err
	}
	if pipelineVersionId != "" || pipelineId != "" {
	    if workflowManifest != "" {
	        return util.NewInvalidInputError("Please either specify a pipeline version or a workflow manifest, not both.")
	    }
	    if pipelineVersionId != "" {
	    	// TODO: shall we verify pipelineId should be parent of this version here?
	    	return nil
	    } else {
	        return ValidatePipelineId(pipelineId)
	    }
	} else if workflowManifest != "" {
	    return ValidateWorkflowManifest(workflowManifest)
	} else {
	    // workflowManifest == "" && pipelineId == "" && pipelineVersionId == ""
	    return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline spec or/and pipeline version.")
	}
```
the main ideas:
* give a name to partial results, so conditions are simpler.
* add some comments to clarify assumptions at different stages of the code.
* use nested if, so that it's clear we didn't forget any situation
* I think it makes more sense to break `ValidatePipelineSpec` into individual functions that either validate workflow manifests, or validates pipeline id.

also, there are similar if else checks in `ValidatePipelineSpec`, we no longer need them when centralizing all the conditions here.

File: backend/src/apiserver/server/util.go [hunk 8]
```diff
@@ -216,9 +216,25 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
 	return nil
 }
 
+func validatePipelineSpecAndCheckPipelineVersionReference(resourceManager *resource.ResourceManager, spec *api.PipelineSpec, resourceReferences []*api.ResourceReference) error {
```
Comment (Bobgy) [inline backend/src/apiserver/server/util.go:219]: what do you think about "validatePipeline" instead?
`pipelineSpec` and `pipelineVersionReference` seems like implementation details.

EDIT: in fact, after reading `ValidatePipelineSpec` code, I think it's simpler overall to add the new logic to `ValidatePipelineSpec` and include resource references as another input argument to `ValidatePipelineSpec`.

File: backend/src/apiserver/server/util.go (hunk 3 above)
Comment (Bobgy) [reply backend/src/apiserver/server/util.go:235]: good catch! You can change methods back to private when you see fit

File: backend/src/apiserver/server/run_server_test.go (hunk 2 above)
Comment (Bobgy) [reply backend/src/apiserver/server/run_server_test.go:459]: This is a good question. I don't have a good answer either.
I usually test all the corner cases in the lowest level UT (e.g. UT of validatePipeline) and only verify e2e integration is correct in UTs that integrate more components (e.g. UT of job/run server).

**Review by Bobgy — COMMENTED** [review]
Thanks @capri-xiyue!
Looks great!

I'm leaving some nit pickings

**Comment by capri-xiyue** [comment]
@Bobgy 
https://github.com/kubeflow/pipelines/pull/4914#discussion_r547051823
Regarding this comment,
my main concern is that usually we don't test private methods(FYI: go/unit-test-practices#public-apis).
Ideally ,we should make validatePipeline() private since it's only used in the package. When we make it private, usually we won't write UT for it. We will write UT for public methods like CreateRunRequest() and CreateJob(). Therefore, we have to duplicate the UT logic for CreateRunRequest() and CreateJob().

File: backend/src/apiserver/server/run_server_test.go (hunk 2 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/run_server_test.go:459]: My main concern is that usually we don't test private methods(FYI: go/unit-test-practices#public-apis).
Ideally ,we should make validatePipeline() private since it's only used in the package. When we make it private, usually we won't write UT for it. We will write UT for public methods like CreateRunRequest() and CreateJob(). Therefore, we have to duplicate the UT logic for CreateRunRequest() and CreateJob().

File: backend/src/apiserver/server/util.go (hunk 8 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/util.go:219]: I prefer to rename it to "validatePipeline" instead of including resource references as another input argument to ValidatePipelineSpec because resource references is not part of PipelineSpec.

Comment (capri-xiyue) [reply backend/src/apiserver/server/util.go:219]: I prefer to rename the method to `validatePipelineSpecAndVersionId`.
If include resource references as another input argument to `ValidatePipelineSpec`, I think it may be kind of confusing because resource references  is not part of PipelineSpec

File: backend/src/apiserver/server/util.go (hunk 7 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/util.go:233]: I think it may be more clear to differentiate the case by workflowmanifest is euqal to "" or not

**Comment by capri-xiyue** [comment]
When I go through the code, the way we write UT does not follow good practice and it is not consistent.
FYI: https://medium.com/@matryer/5-simple-tips-and-tricks-for-writing-unit-tests-in-golang-619653f90742
I will create another issue regarding the clean up of UT

File: backend/src/apiserver/server/run_server_test.go (hunk 2 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/run_server_test.go:459]: By the way, I think it may be better if we put https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/server/util.go under its own package like util and make some util methods public.

File: backend/src/apiserver/server/job_server_test.go (hunk 1 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/job_server_test.go:164]: discussed offline. Cases got covered in UT of validatePipelineSpecAndResourceReferences. Test some cases double in UT of job and run.

File: backend/src/apiserver/server/run_server_test.go (hunk 2 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/run_server_test.go:459]: discussed offline. Will move https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/server/util.go under its own package like util and make some util methods public in another PR.

File: backend/src/apiserver/server/util_test.go (hunk 6 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/util_test.go]: Discussed offline. The util should be moved to another package and some util methods should be public.

File: backend/src/apiserver/server/util.go [hunk 9]
```diff
@@ -216,64 +216,88 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
 	return nil
 }
 
-func ValidatePipelineSpec(resourceManager *resource.ResourceManager, spec *api.PipelineSpec) error {
-	if spec == nil || (spec.GetPipelineId() == "" && spec.GetWorkflowManifest() == "") {
-		return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline ID or workflow manifest.")
+func ValidatePipelineSpecAndResourceReferences(resourceManager *resource.ResourceManager, spec *api.PipelineSpec, resourceReferences []*api.ResourceReference) error {
+	// centralize all the nil checks here
```
Comment (Bobgy) [inline backend/src/apiserver/server/util.go:220]: nit: I guess we can remove this?

File: backend/src/apiserver/server/util.go [hunk 10]
```diff
@@ -216,64 +216,88 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
...
+	pipelineId := spec.GetPipelineId()
```
Comment (Bobgy) [inline backend/src/apiserver/server/util.go:221]: Is the reason directly using `spec.GetPipelineId()` already checks if `spec` is nil?
Good to know, I just learned this is possible today.

File: backend/src/apiserver/server/util.go [hunk 11]
```diff
@@ -216,64 +216,88 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
...
+	if workflowManifest != "" {
+		if pipelineId != "" || pipelineVersionId != "" {
+			return util.NewInvalidInputError("Please don't specify a pipeline version or pipeline ID when you specify a workflow manifest.")
+		}
+		if err := validateWorkflowManifest(spec.GetWorkflowManifest()); err != nil {
+			return err
+		}
+	} else {
+		if pipelineId == "" && pipelineVersionId == "" {
+			return util.NewInvalidInputError("Please specify a pipeline by providing a pipeline spec or/and pipeline version.")
```
Comment (Bobgy) [inline backend/src/apiserver/server/util.go:237]: nit: or workflow manifest

File: backend/src/apiserver/server/util.go [hunk 12]
```diff
@@ -216,64 +216,88 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
...
+		}
+		if err := validatePipelineId(resourceManager, pipelineId); err != nil {
+			return err
+		}
+		if pipelineVersionId != "" {
+			// verify pipelineVersionId exists
+			pipelineVersion, err := resourceManager.GetPipelineVersion(pipelineVersionId)
```
Comment (Bobgy) [inline backend/src/apiserver/server/util.go:244]: discussion: this isn't a problem with this PR, but I think we should keep it in mind and think about a solution.
Right now, handling a single API request may trigger many resource get requests from mysql:
* once in validation
* once in authorization
* once in actually handling the request

But essentially, they are getting the same information.
Any ideas to remove the redundancy?
(maybe let's create a separate issue for this)

File: backend/src/apiserver/server/util.go [hunk 13]
```diff
@@ -216,64 +216,88 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
...
+		if _, err := resourceManager.GetPipeline(pipelineId); err != nil {
+			return util.Wrap(err, "Get pipelineId failed.")
 		}
 	}
-	if spec.GetWorkflowManifest() != "" {
+	return nil
+}
+
+func validateWorkflowManifest(workflowManifest string) error {
+	if workflowManifest != "" {
```
Comment (Bobgy) [inline backend/src/apiserver/server/util.go:282]: nit: although the `workflowManifest == ""` branch may never be reached here

I think it makes more sense to return an error, because when we validate workflow manifest, empty manifest should be an error.

File: backend/src/apiserver/server/util.go [hunk 14]
```diff
@@ -216,64 +216,88 @@ func ValidateExperimentResourceReference(resourceManager *resource.ResourceManag
...
+			}
+			// verify pipelineId should be parent of pipelineVersionId
+			if pipelineId != "" && pipelineVersion.PipelineId != pipelineId {
+				return util.NewInvalidInputError("pipeline ID should be parent of pipeline version.")
+			}
+		}
 	}
-	if spec.GetPipelineId() != "" && spec.GetWorkflowManifest() != "" {
-		return util.NewInvalidInputError("Please either specify a pipeline ID or a workflow manifest, not both.")
+	return validateParameters(spec.GetParameters())
```
Comment (Bobgy) [inline backend/src/apiserver/server/util.go:254]: it seems you also validated parameters at the beginning of this function

File: backend/src/apiserver/resource/resource_manager.go [hunk 15]
```diff
@@ -992,6 +976,23 @@ func (r *ResourceManager) getWorkflowSpecBytesFromPipelineVersion(references []*
 	return []byte(workflow.ToStringForStore()), nil
 }
 
+func getWorkflowSpecManifestBytes(pipelineSpec *api.PipelineSpec, resourceReferences *[]*api.ResourceReference, r *ResourceManager) ([]byte, error) {
+	var workflowSpecManifestBytes []byte
+	if pipelineSpec != nil && pipelineSpec.WorkflowManifest != "" {
```
Comment (Bobgy) [inline backend/src/apiserver/resource/resource_manager.go:981]: nit:
```suggestion
	if pipelineSpec.GetWorkflowManifest() != "" {
```

File: backend/src/apiserver/resource/resource_manager.go [hunk 16]
```diff
@@ -318,18 +318,10 @@ func (r *ResourceManager) CreateRun(apiRun *api.Run) (*model.RunDetail, error) {
...
 	// (1) raw pipeline manifest in pipeline_spec
 	// (2) pipeline version in resource_references
-	// And the latter takes priority over the former
-	var workflowSpecManifestBytes []byte
-	err := ConvertPipelineIdToDefaultPipelineVersion(apiRun.PipelineSpec, &apiRun.ResourceReferences, r)
-	if err != nil {
-		return nil, util.Wrap(err, "Failed to find default version to create run with pipeline id.")
-	}
-	workflowSpecManifestBytes, err = r.getWorkflowSpecBytesFromPipelineVersion(apiRun.GetResourceReferences())
+	// And the latter takes priority over the former when the pipeline manifest is from pipeline_spec.pipeline_id
```
Comment (Bobgy) [inline backend/src/apiserver/resource/resource_manager.go:321]: Shall we add a comment here saying "manifest and pipeline id/version will not exist at the same time, guaranteed by the validation phase"?

**Review by Bobgy — COMMENTED** [review]
/lgtm
The updates LGTM. Only minor comments left

File: backend/src/apiserver/server/util.go (hunk 13 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/util.go:282]: I don't think empty manifest should be an error. For example,  workflow manifest can be empty when we have pipeline id. Same logic used when we validatePipelineId.

File: backend/src/apiserver/server/util.go (hunk 14 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/util.go:254]: done

File: backend/src/apiserver/resource/resource_manager.go (hunk 15 above)
Comment (capri-xiyue) [reply backend/src/apiserver/resource/resource_manager.go:981]: done

File: backend/src/apiserver/resource/resource_manager.go (hunk 16 above)
Comment (capri-xiyue) [reply backend/src/apiserver/resource/resource_manager.go:321]: done

File: backend/src/apiserver/server/util.go (hunk 9 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/util.go:220]: done

File: backend/src/apiserver/server/util.go (hunk 10 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/util.go:221]: Yep. spec.GetPipelineId() already checks if spec is nil or not.

File: backend/src/apiserver/server/util.go (hunk 11 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/util.go:237]: The reason I use pipeline spec here instead of workflow manifest and pipeline id is because pipeline id can coexist with pipelineversion but workflow manifest can't.

File: backend/src/apiserver/server/util.go (hunk 13 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/util.go:278]: Whether empty manifest is valid or not depends on other variables. We won't know it when we validate workflow manifest itself and I don't think we should include the logic when we validate workflow manifest itself.

File: backend/src/apiserver/server/util.go (hunk 11 above)
Comment (Bobgy) [reply backend/src/apiserver/server/util.go:233]: +1, I like your rephrased version

File: backend/src/apiserver/server/util.go (hunk 13 above)
Comment (Bobgy) [reply backend/src/apiserver/server/util.go:278]: I'm thinking about a different definition of what this function does -- "when we have already decided to use workflow manifest, is it valid?". My rationale is that, we've already handled the complexity of when to choose which pipeline source outside. So when we actually call `validateWorkflowManifest`, we know we will be using workflow manifest as pipeline source and empty is not a valid workflow manifest.

The difference is very nuanced and the code path won't be reached, so above is just explaining my thoughts. You can make the decision.

**Comment by Bobgy** [comment]
/lgtm
/approve
Thank you again! This is a really nice clean up of the complex and confusing logic for determining how to get a pipeline!

/hold
to give you a chance to resolve the last comments

File: backend/src/apiserver/server/util.go (hunk 11 above)
Comment (capri-xiyue) [reply backend/src/apiserver/server/util.go]: done

**Review by Bobgy — COMMENTED** [review]
/lgtm
